pkg/apis/awsprovider/v1alpha1/awsclusterproviderconfig_types.go

```diff
@@ -36,15 +36,33 @@ type AWSClusterProviderSpec struct {
 	// SSHKeyName is the name of the ssh key to attach to the bastion host.
 	SSHKeyName string `json:"sshKeyName,omitempty"`
 
-	// CACertificate is a PEM encoded CA Certificate for the control plane nodes.
-	CACertificate []byte `json:"caCertificate,omitempty"`
+	// CAKeyPair is the key pair for ca certs.
+	CAKeyPair KeyPair `json:"caKeyPair,omitempty"`
 
-	// CAPrivateKey is a PEM encoded PKCS1 CA PrivateKey for the control plane nodes.
-	CAPrivateKey []byte `json:"caKey,omitempty"`
+	//EtcdCAKeyPair is the key pair for etcd.
+	EtcdCAKeyPair KeyPair `json:"etcdCAKeyPair,omitempty"`
+
+	// FrontProxyCAKeyPair is the key pair for FrontProxyKeyPair.
+	FrontProxyCAKeyPair KeyPair `json:"frontProxyCAKeyPair,omitempty"`
+
+	// SAKeyPair is the service account key pair.
+	SAKeyPair KeyPair `json:"saKeyPair,omitempty"`
+}
+
+// KeyPair is how operators can supply custom keypairs for kubeadm to use.
+type KeyPair struct {
+	// base64 encoded cert and key
+	Cert []byte `json:"cert"`
+	Key  []byte `json:"key"`
 }
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 func init() {
 	SchemeBuilder.Register(&AWSClusterProviderSpec{})
 }
+
+// HasCertAndKey returns whether a keypair contains cert and key of non-zero length.
+func (kp *KeyPair) HasCertAndKey() bool {
```

Comment:
Given that I see a lot of `!HasCertAndKey()` I wonder if it'd be better to flip this method and call it `IsZero() bool` like `time.Time`.

Reply:
I think `HasCertAndKey` is more explicit in what this method is doing. Also, I want to make this method validate the supplied cert and key are valid crypto materials. I'll do it in a follow-up PR.

Reply:
Got it, we should probably have two separate methods: `IsValid` and `IsZero`. Where `IsValid` calls `IsZero` as one of the validation steps.